### PR TITLE
Snapshots

### PR DESCRIPTION
**Pull request by ivanyu**

### CODE REVIEW

**Comment by fogfish** [comment]
We need to implement following end-points:
1. `/snapshots` list of available snapshots
1. `/snapshots/{id}` replay snapshot identified by `{id}` 
1. `/snapshots/{id}/entities/{key}` retrieve entity value from snapshot 

`HttpdSnapshot.scala` used to be a stub implementation for first end-point.

File: src/main/scala/org/zalando/znap/restapi/HttpdSnapshot.scala [hunk 1]
```diff
@@ -8,25 +8,54 @@
...
-        completeWith[String](implicitly[ToResponseMarshaller[String]]) {
-          commit =>
-            sys.actorSelection("/user/snapshot") ! commit
+  def apiSnapshot(config: Config) = {
+    val targets = config.Targets.map(t => t.id -> t).toMap
+    path("snapshots" / Segment) {
+      (targetId: String) =>
+        get {
+          encodeResponseWith(Gzip) {
+            targets.get(targetId) match {
```
Comment (fogfish) [inline src/main/scala/org/zalando/znap/restapi/HttpdSnapshot.scala]: I would split the long DSL to two functions. There is a guideline:

>
> The first rule of functions is that they should be small. The second rule of functions is that they should be smaller than that. Functions should not be 100 lines long. Functions should hardly ever be 20 lines long.
>  Every function in his program was just two, or three, or four lines long. Each was transparently obvious. Each told a story. And each led you to the next in a compelling order. That’s how short your functions should be!
> 

There is very good guideline. It uses Erlang as example but the consideration is applicable to any FP
http://www.gar1t.com/blog/solving-embarrassingly-obvious-problems-in-erlang.html

File: src/main/scala/org/zalando/znap/restapi/HttpdSnapshotEntity.scala [hunk 2]
```diff
@@ -22,15 +26,35 @@ import scala.concurrent.duration._
...
 
-  def apiSnapshotEntity() = {
     path("snapshots" / Segment / "entities" / Segment) {
-      (snap: String, key: String) => {
+      (targetId: String, key: String) => {
         get {
-          complete {
-            (sys.actorSelection(s"/user/$snap/entity") ? key).mapTo[Option[String]]
+          encodeResponseWith(Gzip) {
+            targets.get(targetId) match {
```
Comment (fogfish) [inline src/main/scala/org/zalando/znap/restapi/HttpdSnapshotEntity.scala]: I would split the long DSL to two functions... See my previous comment

File: src/main/scala/org/zalando/znap/service/SnapshotService.scala [hunk 3]
```diff
@@ -7,27 +7,154 @@
...
+      client,
+      destination.tableName, config.DynamoDB.KVTables.Attributes.Key,
+      scanItemLimit, totalSegments, segment)
+
+    override def receive: Receive = {
+      case Request(n) =>
+        var i = 0L
+        var break = false
+        while (!break && i < n) {
+          try {
```
Comment (fogfish) [inline src/main/scala/org/zalando/znap/service/SnapshotService.scala]: Is it necessary in this model?

File: src/main/scala/org/zalando/znap/service/SnapshotService.scala [hunk 4]
```diff
@@ -7,27 +7,154 @@
...
-    case commit: SnapshotService.Commit =>
-      commit("snapshot")
-      context.stop(self)
+    private val client = {
+      val c = new AmazonDynamoDBClient()
+      c.withEndpoint(destination.uri.toString)
+      c
+    }
+
+    private val itemsIter = new ScannerIterator(
```
Comment (fogfish) [inline src/main/scala/org/zalando/znap/service/SnapshotService.scala]: `Streams` data structure fits the use-case here. The iteration becomes either `fold` or `foreach` and It make code readable

File: src/main/scala/org/zalando/znap/service/SnapshotService.scala (hunk 3 above)
Comment (ivanyu) [inline src/main/scala/org/zalando/znap/service/SnapshotService.scala]: @fogfish yes, it seems so. There is difference (from Streams and Akka HTTP perspective) between a producer actor failing with exception and it calling `onError`. So, to actually call the method on an exception, we have to catch it.

File: src/main/scala/org/zalando/znap/service/SnapshotService.scala (hunk 4 above)
Comment (ivanyu) [inline src/main/scala/org/zalando/znap/service/SnapshotService.scala]: No, it doesn't fit for two reasons.

1. Iteration is controlled from the outside, so `fold` and `foreach` and other combinators aren't applicable.
2. Once evaluated, `Stream` elements stay in memory until the whole stream is GC-ed.

**Pull request merged**